Solve issues with running the benchmark

tatiana opened the pull request:
Fix: #798

Before this PR, the commands (run from `python-sdk/tests/benchmark`):
```
make local
make container
```

Were failing, partially due to the mono-repo refactoring. This PR solves this by also running the benchmark on GCP.

One key component for the success of running the benchmark locally (inside or outside a container) is to make sure the service account credentials are valid. This can be tested by using:
```
gcloud auth activate-service-account --key-file=
```

These were the commands run to confirm if things are working as expected:

# Local run outside of docker
```
GCP_BUCKET=dag-authoring ./run.sh 
```
Which generated:
```
gsutil cat gs://dag-authoring/benchmark/results/results-2022-09-12T11:49:02.ndjson
```

# Local run inside docker
```
GOOGLE_APPLICATION_CREDENTIALS=/Users/al-cht01/Certificates/gcp-astro.json make local
```

# Run inside K8s cluster
```
make setup_gke
make run_job
```

This run is in progress, and it is outputting this file:
```
gsutil cat gs://dag-authoring/benchmark/results/results-2022-09-12T13:03:36.ndjson
```

Which already has the following content:
```
{"duration": 11.076506, "memory_full_info": {"rss": 58396672, "vms": 692084736, "shared": 18092032, "text": 0, "lib": 0, "data": 99639296, "dirty": 0, "uss": 59473920, "pss": 59531264, "swap": 0}, "cpu_time": {"user": 2.5700000000000003, "system": 0.12, "children_user": 0.0, "children_system": 0.0, "iowait": 0.03}, "disk_usage": 196608, "io_counters": [{"read_count": 4030, "write_count": 1155, "read_bytes": 0, "write_bytes": 2945024, "read_chars": 16521137, "write_chars": 2854728}], "dag_id": "load_file_ten_kb_into_snowflake", "execution_date": "2022-09-12 13:04:00.802242+00:00", "revision": "382c4a0", "chunk_size": 1000000}
{"duration": 11.841182, "memory_full_info": {"rss": 46878720, "vms": 89387008, "shared": 7221248, "text": 0, "lib": 0, "data": 39686144, "dirty": 0, "uss": 46845952, "pss": 46899200, "swap": 0}, "cpu_time": {"user": 2.5700000000000003, "system": 0.10999999999999999, "children_user": 0.0, "children_system": 0.0, "iowait": 0.03}, "disk_usage": 4096, "io_counters": [{"read_count": 4142, "write_count": 1157, "read_bytes": 0, "write_bytes": 2945024, "read_chars": 16633338, "write_chars": 2855465}], "dag_id": "load_file_hundred_kb_into_snowflake", "execution_date": "2022-09-12 13:04:20.975850+00:00", "revision": "382c4a0", "chunk_size": 1000000}
{"duration": 11.659376, "memory_full_info": {"rss": 46968832, "vms": 100462592, "shared": 7086080, "text": 0, "lib": 0, "data": 50761728, "dirty": 0, "uss": 46825472, "pss": 46882816, "swap": 0}, "cpu_time": {"user": 2.62, "system": 0.13, "children_user": 0.0, "children_system": 0.0, "iowait": 0.04}, "disk_usage": 8192, "io_counters": [{"read_count": 7836, "write_count": 1157, "read_bytes": 0, "write_bytes": 2945024, "read_chars": 26980928, "write_chars": 2855427}], "dag_id": "load_file_ten_mb_into_snowflake", "execution_date": "2022-09-12 13:04:41.363188+00:00", "revision": "382c4a0", "chunk_size": 1000000}
```

feluelle reviewed (approved):
LGTM

pankajkoti reviewed (commented):
Thanks for the fixes. With this branch, the run script works fine locally but I'm still seeing the below error on GKE:
```
(base) ➜  ~ kubectl -n benchmark logs benchmark-rb478-btljz
standard_init_linux.go:228: exec user process caused: exec format error
```

tatiana commented:
@pankajkoti I just noticed the docker image container was updated - and was able to reproduce your issue if I don't rebuild the image from my computer. Are you using M1? Have you tried building the container & triggering the deployment from a VM / Intel processor? We may have to give some additional parameters to docker when running from M1 - but I don't have one to test right now

pankajkoti commented:
> @pankajkoti I just noticed the docker image container was updated - and was able to reproduce your issue if I don't rebuild the image from my computer. Are you using M1? Have you tried building the container & triggering the deployment from a VM / Intel processor? We may have to give some additional parameters to docker when running from M1 - but I don't have one to test right now

yes @tatiana I am on Mac M1. Probably that could be an issue. I have not tried with VM/Intel. If it works for you, then great. I will use a VM next time then :)

dimberman commented on python-sdk/tests/benchmark/Dockerfile:
```diff
@@ -24,9 +25,11 @@ RUN apt install -y vim
 
+# Install Astro SDK dependencies
 COPY . .
+COPY python-sdk .
```
Maybe it would make more sense to set the WORKDIR to python-sdk here? Is there anything in the base directory we really need?

dimberman commented on python-sdk/tests/benchmark/Makefile:
```diff
@@ -23,23 +23,27 @@ clean:
 	@rm -f ../../unittests.db
 	@rm -f ../../airflow.cfg
 	@rm -f ../../airflow.db
+	@rm -f ../../../unittests.cfg
+	@rm -f ../../../unittests.db
+	@rm -f ../../../airflow.cfg
+	@rm -f ../../../airflow.db
```
This feels odd. I think having two copies of everything leaves more surface area for confusion

tatiana commented:
@pankajkoti as we discussed, let's try to build it using docker buildx (as described in https://blog.jaimyn.dev/how-to-build-multi-architecture-docker-images-on-an-m1-mac/) - but as a separate ticket 🤞 

@dimberman I addressed all your feedback

kaxil commented on python-sdk/tests/benchmark/Makefile:
```diff
@@ -25,29 +25,33 @@ clean:
+container: clean
 	@echo "Building and pushing container $(CONTAINER_REGISTRY)"
-	@gcloud auth configure-docker
+	#@gcloud auth configure-docker
```
Do we not need this anymore?

tatiana replied:
I'll uncomment and improve this, @kaxil , by replacing it with
```
@gcloud auth configure-docker gcr.io
```

By default, this command adds lots of items to `credHelpers` in `~/.docker/config.json` - which slows down docker building. It even prompts the following:

```
WARNING: A long list of credential helpers may cause delays running 'docker build'. We recommend passing the registry name to configure only the registry you are using.
After update, the following will be written to your Docker config file located at [/Users/al-cht01/.docker/config.json]:
 {
  "credHelpers": {
    "gcr.io": "gcloud",
    "us.gcr.io": "gcloud",
    "eu.gcr.io": "gcloud",
    "asia.gcr.io": "gcloud",
    "staging-k8s.gcr.io": "gcloud",
    "marketplace.gcr.io": "gcloud"
  }
}

Do you want to continue (Y/n)?
```

kaxil commented on python-sdk/tests/benchmark/config.json:
```diff
@@ -34,14 +19,6 @@
       "rows": 160,
       "size": "10 KB"
     },
-    {
-      "conn_id": "bigquery",
-      "file_type": "csv",
-      "name": "hundred_kb",
-      "path": "gs://astro-sdk/benchmark/trimmed/tate_britain/artist_data_100kb.csv",
-      "rows": 748,
-      "size": "100 KB"
-    },
```
Why these removals?

tatiana replied:
These were temporary attempts to see the test running on BQ while I was troubleshooting with @utkarsharma2

kaxil commented on python-sdk/tests/benchmark/run.sh:
```diff
@@ -5,7 +5,7 @@ set -v
 
 
-repeat=${1:-1}  # how many times we want to repeat each DAG run (default: 1)
+repeat=${1:-3}  # how many times we want to repeat each DAG run (default: 1)
```
```suggestion
repeat=${1:-3}  # how many times we want to repeat each DAG run (default: 3)
```

kaxil commented on python-sdk/tests/benchmark/Dockerfile:
```diff
@@ -40,25 +43,13 @@ ENV AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True
-
-# Install the Google SDK
-RUN apt install -y curl
-RUN echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key --keyring /usr/share/keyrings/cloud.google.gpg  add - && apt-get update -y && apt-get install google-cloud-cli -y
```
Do we not need to instead the gcloud cli anymore?

tatiana replied:
We do, but these lines were repeated from the top of the file (8-27 in the current commit), I only removed the duplication

tatiana commented on python-sdk/tests/benchmark/Makefile:
```diff
@@ -25,29 +25,33 @@ clean:
 	@echo "Building and pushing container $(CONTAINER_REGISTRY)"
-	@gcloud auth configure-docker
-	@docker build -t benchmark -f ./Dockerfile  ../../ --build-arg=GIT_HASH=`git rev-parse --short HEAD`
+	#@gcloud auth configure-docker
```
Fixed, more info: https://github.com/astronomer/astro-sdk/pull/825#discussion_r969653042
